File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        
        List<Integer> inorder = new ArrayList<>();
        inorderTraversal(root, inorder);
        // System.out.println(inorder);
        
        for(int i=0; i<inorder.size()-1; i++) {
            if(inorder.get(i) >= inorder.get(i+1)) return false;
        }
        return true;
        
    }
    
    public void inorderTraversal(TreeNode root, List<Integer> inorder) {
        if(root == null) return;
        inorderTraversal(root.left, inorder);
        inorder.add(root.val);
        inorderTraversal(root.right, inorder);
    }
}

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    
    const isValid = (root, min, max) => {
        if (root == null) return true;
        if (root.val >= min && root.val <= max) {
            let isLeftBST = isValid(root.left, min, root.val - 1);
            let isRightBST = isValid(root.right, root.val + 1, max);
            return isLeftBST && isRightBST;
        } 
        return false;
    }
    
    return isValid(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
    
};

